### PR TITLE
fix: update ErrorDialog logic and tests

### PR DESCRIPTION
**Pull request by buenos-nachos**
References #9348. Basically, some prep work for the actual feature. I just figured I'd just get these changes out today.

Had to spend a lot of the day debugging an issue with MUI in React Testing Library. Not sure when it was introduced

## Changes
- Updates the tests for `ErrorDialog` so that it no longer spits out cryptic messages about unexpected state changes
- Updates `ErrorDialog` so that the textbox doesn't start nagging you until the textbox loses focus

https://github.com/coder/coder/assets/28937484/9dd6ef07-2733-4f5a-ae92-ba2b956fdb88

### CODE REVIEW

File: site/src/components/Dialogs/DeleteDialog/DeleteDialog.test.tsx [hunk 1]
```diff
@@ -2,6 +2,23 @@ import { screen } from "@testing-library/react";
...
+  // its state to update after a typing event gets  fired, and React Testing
+  // Library isn't able to catch it, making React DOM freak out because an
+  // "unexpected" state change happened. It won't fail the test, but it makes
+  // the console look really scary because it'll spit out a big warning message.
+  // Tried everything under the sun to catch the state changes the proper way,
+  // but the only way to get around it for now might be to manually make React
+  // DOM aware of the changes
+
+  // eslint-disable-next-line testing-library/no-unnecessary-act -- have to make sure state updates don't slip through cracks
+  return act(() => userEvent.type(inputElement, text));
```
Comment (Parkreiner) [inline site/src/components/Dialogs/DeleteDialog/DeleteDialog.test.tsx:20]: To be clear, this issue existed with the previous version of the component, too – it's just that the test never dealt with the warning messages getting spat out

Going to see if it's worth opening a bug report

Comment (Parkreiner) [reply site/src/components/Dialogs/DeleteDialog/DeleteDialog.test.tsx:20]: The logic in the tests themselves doesn't change

Comment (Kira-Pilot) [reply site/src/components/Dialogs/DeleteDialog/DeleteDialog.test.tsx:20]: Maybe the first reasonable use case for `act` I've seen in my career :)

**Review by Kira-Pilot — APPROVED** [review]
Thanks for working on this - I bet that test was a pita

**Pull request merged**